src/EditorFeatures/DiagnosticsTestUtilities/Diagnostics/AbstractUserDiagnosticTest.cs

```diff
@@ -74,6 +76,42 @@ public abstract partial class AbstractUserDiagnosticTest : AbstractCodeActionOrU
             return dxs;
         }
 
+        protected void AddAnalyzerToWorkspace(Workspace workspace, DiagnosticAnalyzer analyzer, TestParameters parameters)
+        {
+            AnalyzerReference[] analyzeReferences;
+            if (analyzer != null)
+            {
+                Contract.ThrowIfTrue(parameters.runProviderOutOfProc, $"Out-of-proc testing is not supported since {analyzer} can't be serialized.");
+
+                analyzeReferences = new[] { new AnalyzerImageReference(ImmutableArray.Create(analyzer)) };
+            }
+            else
+            {
+                // create a serializable analyzer reference:
+                analyzeReferences = new[]
+                {
+                    new AnalyzerFileReference(DiagnosticExtensions.GetCompilerDiagnosticAnalyzer(LanguageNames.CSharp).GetType().Assembly.Location, FromFileLoader.Instance),
+                    new AnalyzerFileReference(DiagnosticExtensions.GetCompilerDiagnosticAnalyzer(LanguageNames.VisualBasic).GetType().Assembly.Location, FromFileLoader.Instance)
+                };
+            }
+
+            workspace.TryApplyChanges(workspace.CurrentSolution.WithAnalyzerReferences(analyzeReferences));
+        }
+
+        private class FromFileLoader : IAnalyzerAssemblyLoader
```

Comment:
Can we add a common type in Rosly.Test.Utilities project, say `TestAnalyzerAssemblyLoader`, with this implementation and avoid having clones of `FromFileLoader` in each test file?

Reply:
Yes, in follow up review. This would require many other places to change.